llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp

```diff
@@ -2485,6 +2490,197 @@ bool ObjCARCOpt::run(Function &F, AAResults &AA) {
+            }
+          }
+
+          if (!LeadsToExpectedPop) {
+            SafeToOptimize = false;
+            LLVM_DEBUG(dbgs() << "Unsafe to optimize: push has unexpected use: "
+                              << *U << "\n");
+            break;
+          }
+        }
+
+        if (!HasAutoreleaseInScope && SafeToOptimize) {
+          LLVM_DEBUG(dbgs() << "Eliminating empty autorelease pool pair: "
+                            << *PendingPush << " and " << *Pop << "\n");
+
+          // Store the pair for careful deletion later
+          EmptyPoolPairs.push_back({PendingPush, Pop});
+
+          ++NumNoops;
+        }
+        break;
+      }
+      case ARCInstKind::CallOrUser:
+      case ARCInstKind::Call:
+      case ARCInstKind::Autorelease:
+      case ARCInstKind::AutoreleaseRV: {
+        // Track that we have autorelease calls in the current pool scope
+        if (!PoolStack.empty()) {
+          PoolStack.back().second = true; // Set has_autorelease_in_scope = true
+          LLVM_DEBUG(
+              dbgs()
+              << "Found autorelease or potiential autorelease in pool scope: "
+              << Inst << "\n");
+        }
+        break;
+      }
+
+      // Enumerate all remaining ARCInstKind cases explicitly
+      case ARCInstKind::Retain:
+      case ARCInstKind::RetainRV:
+      case ARCInstKind::UnsafeClaimRV:
+      case ARCInstKind::RetainBlock:
+      case ARCInstKind::Release:
+      case ARCInstKind::NoopCast:
+      case ARCInstKind::FusedRetainAutorelease:
+      case ARCInstKind::FusedRetainAutoreleaseRV:
+      case ARCInstKind::LoadWeakRetained:
+      case ARCInstKind::StoreWeak:
+      case ARCInstKind::InitWeak:
+      case ARCInstKind::LoadWeak:
+      case ARCInstKind::MoveWeak:
+      case ARCInstKind::CopyWeak:
+      case ARCInstKind::DestroyWeak:
+      case ARCInstKind::StoreStrong:
+      case ARCInstKind::IntrinsicUser:
+      case ARCInstKind::User:
+      case ARCInstKind::None:
+        // These instruction kinds don't affect autorelease pool optimization
+        break;
+      }
+    }
+  }
+
+  // Handle empty pool pairs carefully to avoid use-after-delete
+  SmallVector<CallInst *, 8> DeadInsts;
+  for (auto &Pair : EmptyPoolPairs) {
+    CallInst *Push = Pair.first;
+    CallInst *Pop = Pair.second;
+
+    // Replace all uses of push with poison before deletion
+    Push->replaceAllUsesWith(PoisonValue::get(Push->getType()));
+
+    LLVM_DEBUG(dbgs() << "Erasing empty pool pair: " << *Push << " and " << *Pop
+                      << "\n");
+    DeadInsts.push_back(Pop);
+    DeadInsts.push_back(Push);
+  }
+
+  // Remove the pairs
+  if (!DeadInsts.empty()) {
+    Changed = true;
+    for (CallInst *DeadInst : DeadInsts) {
+      LLVM_DEBUG(dbgs() << "Erasing dead instruction: " << *DeadInst << "\n");
+      DeadInst->eraseFromParent();
+    }
+
+    ORE.emit([&]() {
+      return OptimizationRemark(DEBUG_TYPE, "AutoreleasePoolElimination", &F)
```

Comment:
If you do one remark per `EmptyPoolPair`, the debuginfo tying it to a particular source location will be much more useful in the opt-viewer (and related tools) to help users visualize where they're coming from. Plus, `llvm-remarkutil`'s `count` tool can do the per-function aggregation after the fact.